test/postControllerTest.js

```diff
@@ -0,0 +1,96 @@
+/* global it, describe */
+const express = require('express');
+const request = require('supertest');
+const postController = require('../controllers/postController');
+const sinon = require('sinon');
+const postService = require('../services/postService');
+
+const app = express();
+
+app.set('view engine', 'ejs');
+app.use(express.static('../public'));
+
+app.use('', postController);
+const server = app.listen(3000);
+
+
+describe('postController', () => {
+  describe('GET /', () => {
+    it('should respond with homepage', (done) => {
+      const stub = sinon.stub(postService, 'getAllPosts').resolves('expected');
+      request(app)
+        .get('/')
+        .expect(200)
+        .expect('Content-Type', 'text/html; charset=utf-8')
+        .expect(/My First Blog/)
+        .end((err) => {
+          stub.restore();
+          if (err) return done(err);
+          return done();
+        });
+    });
+  });
+  describe('GET /posts', () => {
+    it('should respond with homepage', (done) => {
+      const stub = sinon.stub(postService, 'getAllPosts').resolves('expected');
+      request(app)
+        .get('/posts')
+        .expect(200)
+        .expect('Content-Type', 'text/html; charset=utf-8')
+        .expect(/My First Blog/)
+        .end((err) => {
+          stub.restore();
+          if (err) return done(err);
+          return done();
+        });
+    });
+  });
+  describe('GET /posts/:post_id', () => {
+    it('should respond with post detail', (done) => {
+      const stub = sinon.stub(postService, 'getOnePost').resolves('expected');
```

Comment:
[Mocha hook](https://mochajs.org/#hooks)

So it looks like every single test here will stub out postService's methods, perhaps with different behaviors. You can move the stubbing and restoring into mocha hooks and keep the mock behavior within your tests

Reply:
I see how I can add stub.restore() to a afterEach() mocha hook. However I don't see how I can move the stubbing to a hook, besides creating multiple stubs for each postService method and adding them all to a before() hook. Which doesn't seem to improve readability of my tests.

Reply:
Why doesn't that improve readabiliy? That'll cause `sinon.stub(postService, 'getAllPosts')` to only appear once, instead of twice.

Also, why choose `before` over `beforeEach`?

Reply:
Right, beforeEach is the way to go. I thought readability would be better because people would be able to see which stubs related with which tests, but I guess that's unnecessary as people can see the postService method being tested. However is stubbing expensive? Adding all stubs to beforeEach would mean I'm making stubs even when a specific test may not require it.

Reply:
> postService method being tested

What do you mean? You aren't testing `postService`, you are testing `postController`

[Premature optimization is the root of all evil](http://wiki.c2.com/?PrematureOptimization)

Reply:
Wrong word usage. I meant postService method being stubbed. Right, premature optimization is bad. But I still don't know if moving all the stubs to a before() mocha hook is worth having to perform unnecessary stubs for tests that don't require them. I'm not doubting you, I just genuinely don't know.

Reply:
Ok, you don't know. But do you need to know? Are the tests running slow enough to give this more though?

Reply:
There are so many little things that to me have no discernible effect on how my code runs that end up mattering more than I thought. But I see from your comments that it doesn't matter.

Reply:
This is true, performance might just seem like another aspect similar to readability or naming conventions, but in most scenarios, performance is not something to be worried about until needed